README.md
# Chess Core Lite
A lightweight chess library for frontend chess apps.

# Installation
```
    npm install chess-core-lite
```

# v1.1.1 (3 Jul 2022)

## Changes
- Bug fix: assertPiece and assertSquare not defined

# Features
- Complete chess gameplay support
- Scoreboard (List of moves made)
- Checkmate move finder
- PGN support
- FEN support
- Lightweight module


# Getting Started

```javascript
    import {
        Chess, ChessResult,findMate,
        toFEN, toPGN, fromPGN, newPGNObject, } from 'chess-core-lite';
    
    // New game with default fen
    const game = new Chess(); 

    // New game with custom fen
    const fen = '8/2q2P2/p1R5/1Q1p4/2p3pK/1P1R4/3P1p1P/k6N w - - 0 1';
    const game2 = new Chess(fen);

    // make a move by SAN
    game.moveSAN('e4');
    // make a move by position
    game.move(75, 55);
    // undo move
    game.undoMove();
    // goto nth move
    game.goto(1); // go to 1st move
    game.goto(-1); // go to last move
    // NOTE: if a move made on nth state then all moves made thereafter are cleared.

    // get board state
    const state = game.currentState;
    const lastState = game.lastState; // returns last state irrespective of current state user goes.

    // List the moves made
    const moves = game.score;

    // Check the states of game
    const isGameActive = game.result === ChessResult.GAME_ON;
    game.isCheck
    game.isCheckmate
    

    // Get FEN
    const fen = toFEN(game.currentState);

    // Get PGN
    const pgnObject = newPGNObject(game.score);
    const pgnObject['Event'] = 'Bob v Marley';
    const pgnObject['Date'] = '10-01-21';
    const pgn = toPGN(pgnObject);

    // Load PGN
    const pgnObjects = fromPGN('[Event "BvS"]\n1. e4 e5 2. Nc3 Nf6');
    game.load(pgnObjects[0].score);

    // Find Checkmate move
    const mates = findMate(game.currentState);
    if(mates.length) game.move(...mates[0]); // Do the checkmate move
```

## API

> API docs coming soon

### Chess (class)
- result : ChessResult
- score : string[]
- currentState: BoardState
- lastState: BoardState
- board: Board
- moves: Moves
- isCheck : boolean 
- isCheckmate : boolean 
- isStalemate : boolean 
- isThreefold : boolean 
- isFiftyMove : boolean
- isInsufficient : boolean
- load(score: string[], [fen: string]) : boolean
- move(from: number, to: number, [promotion: string]) : boolean
- moveSAN(san: string) : boolean
- undoMove() : void
- claimDraw() : ChessResult
- reset(): void

### ChessResult (Enum)
- GAME_ON = 0
- WHITE_WON = 1
- BLACK_WON = 10
- STALEMATE = 100
- THREEFOLD_DRAW = 102
- FIFTYMOVE_DRAW = 103
- INSUFFICIENT_DRAW = 104
- CLAIMED_DRAW = 105

### Other Types
Board alias string: a string refence to positions of board

Moves alias object: containing list of legal moves for the particular board state.



## Contribute

```
    git clone https://github.com/TishanV/chess-core.git
    cd chess-core
    npm install
```

**Prefer Typescript**

> Developer guide coming soon.

CHANGELOG.md
# 1.1.1 (3 Jan 2022)
- Bug fix: assertPiece and assertSquare not defined

# 1.1.0 (31 Jan 2022)

## New features
- __listVulnerables__: list piece positions being attacked by opponent.
- __listCaptures__: list piece postitions where opponent pieces are in capture.

## Changes
-  Module system changed to common js from es6 to support node projects.

src/chess.ts
import {
  BoardPosition,
  BoardState,
  PromotionPiece,
  Moves,
  Board,
} from "./types";
import { sameColorPos } from "./core/utils";
import { fenToBoardState } from "./core/boardState";
import { doMove } from "./core/moveable";
import { promotionOf, toCordsFromSAN, toSANMove } from "./san";
import { validateFEN } from "./fen";

export enum ChessResult {
  GAME_ON = 0,
  WHITE_WON = 10,
  BLACK_WON = 100,
  STALEMATE = 101,
  THREEFOLD_DRAW = 102,
  FIFTYMOVE_DRAW = 103,
  INSUFFICIENT_DRAW = 104,
  CLAIMED_DRAW = 105,
}

const defaultFEN: string =
  "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";

export class Chess {
  private _states: BoardState[];
  private _current: number;
  private _score: string[];
  private _result: ChessResult;

  constructor(fen: string = defaultFEN) {
    if (fen !== defaultFEN) fen = validateFEN(fen) ? fen : defaultFEN;
    this._states = [fenToBoardState(fen)];
    this._current = 0;
    this._score = ["-"];
    this._result = ChessResult.GAME_ON;
  }

  public get result(): ChessResult {
    return this._result;
  }

  public get score(): string[] {
    return this._score.slice(1);
  }

  public goto(i: number) {
    if (i < 0) i = this._states.length + i;
    if (i < 0 || i >= this._states.length) return;
    this._current = i;
  }

  public get currentState(): BoardState {
    return this._states[this._current];
  }

  public get lastState(): BoardState {
    return this._states[this._states.length - 1];
  }

  public get board(): Board {
    return this._states[this._current].board;
  }

  public get moves(): Moves {
    return this._states[this._current].moves;
  }

  public claimDraw(): ChessResult {
    return (this._result = ChessResult.CLAIMED_DRAW);
  }

  public load(score: string[], fen = defaultFEN) {
    this.reset();
    this._states = [fenToBoardState(fen)];
    return score.every((san) => this.moveSAN(san));
  }

  public moveSAN(san: string) {
    const cords = toCordsFromSAN(san, this.currentState);
    return cords ? this.move(...cords, promotionOf(san)) : false;
  }

  public move(
    from: BoardPosition,
    to: BoardPosition,
    promotion: PromotionPiece = "Q"
  ) {
    if (this._result !== ChessResult.GAME_ON) return false;
    const newState = doMove(this._states[this._current], [from, to], promotion);
    if (newState !== undefined) {
      if (this._current !== this._states.length - 1) {
        this._states.splice(this._current + 1);
        this._score.splice(this._current + 1);
      }
      this._score.push(toSANMove([from, to], this.currentState, newState));
      this._states.push(newState);
      this._current = this._states.length - 1;
      this.resultCheck();
      return true;
    }
    return false;
  }

  public undoMove() {
    this._states.length > 1
      ? this._states.pop() &&
        this._score.pop() &&
        (this._current = this._states.length - 1)
      : null;
  }

  public reset() {
    this._states.length = 1;
    this._score.length = 1;
    this._current = 0;
    this._result = ChessResult.GAME_ON;
  }

  public get isCheck() {
    return this.currentState.checkline.length > 0;
  }

  private get noMoves() {
    return Object.values(this.currentState.moves).every((e) => !e.length);
  }

  public get isCheckmate() {
    return this.isCheck && this.noMoves;
  }

  public get isStalemate() {
    return !this.isCheck && this.noMoves;
  }

  public get isFiftyMove() {
    return this.currentState.halfMove >= 50;
  }

  public get isThreefold() {
    return (
      this._states.length -
        Array.from(
          new Set(
            this._states.map(
              (e) => `${e.board} ${e.player} ${e.castle} ${e.enpassant}`
            )
          )
        ).length >=
      2
    );
  }

  public get isInsufficient() {
    return (
      (!this.board.match(/[qQrRpP]/g) &&
        (this.board.match(/[bBnN]/g)?.length || 0) < 2) ||
      (this.board.match(/[bBnN]/g)?.length === 2 &&
        this.board.match(/[b]/g)?.length === 1 &&
        this.board.match(/[B]/g)?.length === 1 &&
        sameColorPos(this.board.indexOf("b"), this.board.indexOf("B")))
    );
  }

  private resultCheck() {
    if (this.isCheckmate)
      return (this._result =
        this.currentState.player === "b"
          ? ChessResult.WHITE_WON
          : ChessResult.BLACK_WON);
    if (this.isStalemate) return (this._result = ChessResult.STALEMATE);
    if (this.isThreefold) return (this._result = ChessResult.THREEFOLD_DRAW);
    if (this.isFiftyMove) return (this._result = ChessResult.FIFTYMOVE_DRAW);
    if (this.isInsufficient)
      return (this._result = ChessResult.INSUFFICIENT_DRAW);
  }
}
